# SOS Tutorial #

The purpose of this tutorial is to give a quick start into working with SOS, highlight some of its features, and explain the rationale behind some of its differences and design decisions compared to other version control systems (VCS).

## Getting started ##
First, you need to install SOS and make it available in your system path. This is best done using the `pip` (or `pip3`) command of your Python distribution. If you don't have Python installed, do so using your system's package manager or by downloading and installing it from `https://www.python.org/downloads/` or, e.g., `https://conda.io/miniconda.html`.

Installing SOS through pip works like the following code, but take into account
- the `!` is not part of what you would write on the command line;
- the installation might require admin rights;
- reload this notebook after successful installation):

In [1]:
!pip install sos-vcs

  Using cached https://files.pythonhosted.org/packages/a2/46/fd8d902e8001f4c7282b8822aa2beecb1971f11ca0f6f4862b74162451fd/sos-vcs-2018.1624.2412.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-6mANOM/sos-vcs/setup.py", line 76, in <module>
        import sos.version
      File "sos/__init__.py", line 8, in <module>
        from sos import *
      File "sos/sos.py", line 62
        def loadConfig() -> 'configr.Configr':  # Accessor when using defaults only  # line 34
                         ^
    SyntaxError: invalid syntax
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-install-6mANOM/sos-vcs/
You are using pip version 18.1, however version 19.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


This will install SOS and all its required software packages into your global Python distribution. If you want to test SOS first in a virtual environment, make sure your use a virtual enviroment via [pipenv](https://docs.pipenv.org) or have a `conda`-supporting Python distribution with an activated virtual Python environment and run the `pip` command from there.

Once installed, confirm that everything is fine via

In [40]:
try: pwd = os.getcwd()  # this code is only required in case you run this notebook several times
except: os.chdir("..")

!sos version

          d8  ad88888ba    ,ad8888ba,    ad88888ba           d8     
        ,8P' d8"     "8b  d8"'    `"8b  d8"     "8b        ,8P'     
       d8"   Y8,         d8'        `8b Y8,               d8"       
     ,8P'    `Y8aaaaa,   88          88 `Y8aaaaa,       ,8P'        
    d8"        `"""""8b, 88          88   `"""""8b,    d8"          
  ,8P'               `8b Y8,        ,8P         `8b  ,8P'           
 d8"         Y8a     a8P  Y8a.    .a8P  Y8a     a8P d8"             
8P'           "Y88888P"    `"Y8888Y"'    "Y88888P" 8P'              

/SOS/ Subversion Offline Solution V1.6.9 (C) Arne Bachmann (PyPI: 2019.1210.2731-v1.6.0-88-gaff3ab9)


## First steps ##
Say you have a project you work on, but you are either **a)** using a means of transport without (affordable) internet access, **b)** experiencing a power or network outage, **c)** in need of VCS functionality for something not (yet) part of a repository or checkout.

SOS will support you in any case.

For the sake of this tutorial, we assume you are working in a local Git checkout (could also be a SVN checkout or any other folder you'd like to experiment on).

As I don't know any public SVN platforms to checkout from for this tutorial, **download** and **unzip** or **clone** a random project from Github instead, e.g., [this project's archive](https://github.com/ArneBachmann/realestate-sunamount/archive/master.zip), which is a little project of mine written in the Coconut programming language.

Let's start!

Imagine you are taking a train without network connectivity for some reason and want to start working on the (just downloaded) RESE project to improve some code parts. In the shell type this:

In [41]:
!git clone https://github.com/ArneBachmann/realestate-sunamount.git
!cd realestate-sunamount ; sos offline
!sos config set useColorOutput yes --quiet

Cloning into 'realestate-sunamount'...
remote: Enumerating objects: 119, done.
remote: Total 119 (delta 0), reused 0 (delta 0), pack-reused 119
Receiving objects: 100% (119/119), 56.25 KiB | 606.00 KiB/s, done.
Resolving deltas: 100% (72/72), done.
          d8  ad88888ba    ,ad8888ba,    ad88888ba           d8     
        ,8P' d8"     "8b  d8"'    `"8b  d8"     "8b        ,8P'     
       d8"   Y8,         d8'        `8b Y8,               d8"       
     ,8P'    `Y8aaaaa,   88          88 `Y8aaaaa,       ,8P'        
    d8"        `"""""8b, 88          88   `"""""8b,    d8"          
  ,8P'               `8b Y8,        ,8P         `8b  ,8P'           
 d8"         Y8a     a8P  Y8a.    .a8P  Y8a     a8P d8"             
8P'           "Y88888P"    `"Y8888Y"'    "Y88888P" 8P'              
ADD ./.gitignore  (1157 bytes)                                            
ADD ./LICENSE  (16.33 KiB)
ADD ./README.md  (1.34 KiB)
ADD ./__coconut__.py  (26.10 KiB)
ADD ./example.coco  (3.54 KiB)
ADD 

By "going offline" inside the repository checkout (or any other folder), a metadata sub-folder `.sos` will be created that contains a baseline copy of your project and a record of all the changes you track.
You can immediately start hacking away and not worry about changing anything beyond repair, as you can always revert your changes to where you started from when you went `sos offline` or did your last `sos commit`.

You may also use the command line switches `--compress` and `--strict`:
- `--compress` enables data compression inside the SOS internals, which is slower but uses less data on the file system,
- `--strict` uses hashes on the entire file contents to detect changes instead of relying on size and timestamp only, which is slower but more accurate

OK, let's start working. You open your editor and edit a source file inside the checked out folder `realestate-sunamount`, e.g. `make.bat` from:

```Powershell
@echo off
echo Compiling
...
```

to this:

```Powershell
@echo off
echo Compiling...
...
```

(note the added dots after *Compiling*), or run the next Python cell to do it:

In [42]:
import os
os.chdir("realestate-sunamount")
with open("make.bat", "r") as fd: data = fd.read().split("\n")
data[0] += "..."  # add "..." after "Compiling"
with open("make.bat", "w") as fd: fd.write("\n".join(data))
print("Make.bat was updated.")

Make.bat was updated.


Satisfied with your improvement, you want to commit your change(s) to the local offline SOS history to make a snapshot of your work that you can later refer to or revert to.

But just before committing your change just yet, you'd like to see a summary of your changes to confirm that everything is in order:

In [43]:
!sos status

/SOS/ Offline repository status
Repository root:     /home/ash/Desktop/all/projects/sos/docs/realestate-sunamount
Underlying VCS root: /home/ash/Desktop/all/projects/sos/docs/realestate-sunamount
Underlying VCS type: git
Installation path:   /home/ash/Desktop/all/projects/sos
Current SOS version: 2019.1210.2731-v1.6.0-88-gaff3ab9
At creation version: 2019.1210.2731-v1.6.0-88-gaff3ab9
Metadata format:     2
Content checking:    size & timestamp
Data compression:    deactivated
Repository mode:     simple
Number of branches:  1
✖ File tree has changes
  * b0 'master' @2019-02-10 17:53:39 (in sync) with 1 commits (median 14 each), using 0.13 MiB (+1.805% SOS overhead). Last comment: 'Offline repository created on 2019-02-10 17:53:39'
[EXIT]


or, if you prefer a less Git-like and more Fossil-like command:

In [44]:
!sos config set useChangesCommand yes --quiet
!sos changes

MOD ./make.bat
[EXIT]


The `sos config` command defines a user-global SOS configuration setting (unless using `--local` which would apply to the current checkout only, in contrast to what Git does via `--global`).

The `useChangesCommand` setting modifies the behavior of the `status` command to show the checkout stats instead of the repository status (which otherwise would be shown using `status --repo[sitory]`).

Ok, so we know that `make.bat` has been modified as we wanted. Let's display the changes in more detail:

In [45]:
!sos diff

DIF ./make.bat  <LF>
old 0 |@echo off|
now 0 |@echo off...|

[EXIT]


which tells us exactly on what line what was changed from old to new.
The pipe symbols mark the beginning and end of each displayed line, so you can also notice trailing whitespaces.

It's now time to commit our change into the offline repository's history.
SOS was designed with minimal barriers in mind, therefore you can simply do:

In [46]:
!sos commit

MOD ./make.bat
Processing speed was 96.00 kiB/s.
/SOS/ Created new revision r01 (+00/-00/±01/⇌00) summing 507 bytes in 1 files (33.93% SOS overhead)
[EXIT]


or use the shorter command versions `sos ci` and `sos com`.

This works similar to `git add --all && git commit -m ".."` or `svn ci -m ".."` as it mirrors all (not ignored) file tree changes into the repository.

The output is a condensed summary of changes.
It shows the created revision's number `01`, the commit message, the number of added, removed, modified and moved files, the number of bytes and files processed, and the percentage of metadata overhead occupied by SOS to manage the offline repository, which is usually less than one percent.

You can also specify a commit message, and specify which files to commit or not exclude:

In [47]:
!sos commit "First change" --only "*.bat" --except "*.cmd"

[EXIT Nothing to commit.]


To get an overview of the recent SOS commit history:

In [48]:
!sos log

/SOS/ Offline commit history of branch 'master'
    r0 @2019-02-10 17:53:39 (+14/-00/±00/T00) |Offline repository created on 2019-02-10 17:53:39|
  ➙ r1 @2019-02-10 17:53:51 (+00/-00/±01/T01) ||
[EXIT]


Or even the extended `log` command, which also shows the files modifed in each commit:

In [49]:
!sos log --changes

/SOS/ Offline commit history of branch 'master'
    r0 @2019-02-10 17:53:39 (+14/-00/±00/T00) |Offline repository created on 2019-02-10 17:53:39|
ADD ./.gitignore  (-)
ADD ./LICENSE  (-)
ADD ./README.md  (-)
ADD ./__coconut__.py  (-)
ADD ./example.coco  (-)
ADD ./example.py  (-)
ADD ./make.bat  (-)
ADD ./make.sh  (-)
ADD ./setup.py  (-)
ADD ./sunamount.md  (-)
ADD rese/__coconut__.py  (-)
ADD rese/__init__.py  (-)
ADD rese/sunamount.coco  (-)
ADD rese/sunamount.py  (-)
  ➙ r1 @2019-02-10 17:53:51 (+00/-00/±01/T01) ||
MOD ./make.bat
[EXIT]


You can work this way as long as you want and record regular snapshots of your work until you want to go back online.

This is the main purpose of SOS: record your changes effortlessly over time while during offline periods to allow reverting in case of mistakes. SOS also allows you to switch between branches to test out different ideas and separate change sets.

Once you regain online access you can try the `sos online` command:

In [50]:
!sos online

[EXIT There are still unsynchronized (modified) branches.]
Use 'sos log' to list them.
Use 'sos commit' and 'sos switch' to commit out-of-sync branches to your VCS before leaving offline mode.
Use 'sos online --force' to erase all aggregated offline revisions without further action.


This means that SOS won't let you go back online (which simply removes the entire offline repository !), until you have secured all your offline work back into the underlying VCS (if any).

You can therefore also check the status of the repository and all its branches to check if any offline changes are not yet considered after regaining connectivity:

In [51]:
!sos status --repo

/SOS/ Offline repository status
Repository root:     /home/ash/Desktop/all/projects/sos/docs/realestate-sunamount
Underlying VCS root: /home/ash/Desktop/all/projects/sos/docs/realestate-sunamount
Underlying VCS type: git
Installation path:   /home/ash/Desktop/all/projects/sos
Current SOS version: 2019.1210.2731-v1.6.0-88-gaff3ab9
At creation version: 2019.1210.2731-v1.6.0-88-gaff3ab9
Metadata format:     2
Content checking:    size & timestamp
Data compression:    deactivated
Repository mode:     simple
Number of branches:  1
✔ File tree is unchanged                                             
  * b0 'master' @2019-02-10 17:53:39 (modified) with 2 commits (median 7 each), using 0.13 MiB (+1.994% SOS overhead)
[EXIT]


Here you see (after some repository internals), that the only existing branch is marked as **modified**, which means it has been made changes to since you went offline and wasn't committed to the underlying VCS yet, i.e., using the `sos publish` command.

Make sure you don't forget to secure your changes to the underlying VCS you went offline from before issuing `sos online` or `sos online --force` to remove the offline repository. You can even use SOS to issue the underlying `commit` commandt, using `sos commit --vcs -m "Message"` which calls the underlying VCS and clears the SOS dirty flag.
## Branching ##

This section explains how to rollback changes and how to work on different (feature) branches while being offline.

Let's say we suddenly get the inspiration to work on another idea, but want to pick up work were we left.
This is were branching comes into play:

In [52]:
!sos branch --fast

ADD ./.gitignore  (1157 bytes)
ADD ./LICENSE  (16.33 KiB)
ADD ./README.md  (1.34 KiB)
ADD ./__coconut__.py  (26.10 KiB)
ADD ./example.coco  (3.54 KiB)
ADD ./example.py  (4.58 KiB)
ADD ./make.bat  (335 bytes)
ADD ./make.sh  (364 bytes)
ADD ./setup.py  (2.28 KiB)
ADD ./sunamount.md  (12.18 KiB)
ADD rese/__coconut__.py  (26.10 KiB)
ADD rese/__init__.py  (0 bytes)
ADD rese/sunamount.coco  (15.63 KiB)
ADD rese/sunamount.py  (21.95 KiB)
Processing speed was 21.74 MiB/s.
/SOS/ Switched to new unnamed branch b1
[EXIT]


A new branch was created, and you have been switched to work on it. Ignore the `--fast` for the moment, it is simply faster if you switch immediately after a commit.

Check the branches by typing:

In [53]:
!sos status

/SOS/ Offline repository status
Repository root:     /home/ash/Desktop/all/projects/sos/docs/realestate-sunamount
Underlying VCS root: /home/ash/Desktop/all/projects/sos/docs/realestate-sunamount
Underlying VCS type: git
Installation path:   /home/ash/Desktop/all/projects/sos
Current SOS version: 2019.1210.2731-v1.6.0-88-gaff3ab9
At creation version: 2019.1210.2731-v1.6.0-88-gaff3ab9
Metadata format:     2
Content checking:    size & timestamp
Data compression:    deactivated
Repository mode:     simple
Number of branches:  2
✔ File tree is unchanged
    b0 'master' @2019-02-10 17:53:39 (modified) with 2 commits (median 7 each), using 0.13 MiB (+1.994% SOS overhead)
  * b1          @2019-02-10 17:53:59 (modified) with 1 commits (median 14 each), using 0.13 MiB (+1.796% SOS overhead). Last comment: 'Branched from file tree after b0/r01'
[EXIT]


The asterisk `*` shows you the currently active (unnamed) brach.

We now make a change in `rese/sunamount.coco` by moving the conditional `import sys` statement from the bottom to the top of the file (use the following code to let the notebook do it for you):

In [54]:
with open("rese/sunamount.coco", "r") as fd: lines = fd.read().replace("\r\n", "\n").split("\n")
lines = lines[:8] + ["import sys"] + [line for line in lines[8:] if line.strip() != "import sys"]
with open("rese/sunamount.coco", "w") as fd: fd.write("\n".join(lines))
!echo
!sos diff


DIF rese/sunamount.coco  <LF>
+++ 008 |import sys|

--- 274 |  import sys|

[EXIT]


If you hadn't had SOS installed earlier, SOS will treat the Coconut source file as binary.
We will tell SOS to treat `*.coco` files globally as textual files and repeat the `sos diff` command:

In [55]:
!sos config add texttype "*.coco" --local
!echo
!sos diff

[EXIT Value already contained, nothing to do.]

DIF rese/sunamount.coco  <LF>
+++ 008 |import sys|

--- 274 |  import sys|

[EXIT]


OK, the import statement has been removed from line 274 and inserted before line 8 (0-based) or inserted after line 8 (1-based).

We can now commit the changes we had in mind and go back to our original branch to continue worke there:

In [56]:
!sos commit "Moved import statement"
!echo
!sos log
!echo
!sos status

MOD rese/sunamount.coco
Processing speed was 4.10 MiB/s.
/SOS/ Created new revision r01 'Moved import statement' (+00/-00/±01/⇌00) summing 15.81 KiB in 1 files (1.13% SOS overhead)
[EXIT]

/SOS/ Offline commit history of branch None
    r0 @2019-02-10 17:53:59 (+14/-00/±00/T00) |Branched from file tree after b0/r01|
  ➙ r1 @2019-02-10 17:54:06 (+00/-00/±01/T01) |Moved import statement|
[EXIT]

/SOS/ Offline repository status
Repository root:     /home/ash/Desktop/all/projects/sos/docs/realestate-sunamount
Underlying VCS root: /home/ash/Desktop/all/projects/sos/docs/realestate-sunamount
Underlying VCS type: git
Installation path:   /home/ash/Desktop/all/projects/sos
Current SOS version: 2019.1210.2731-v1.6.0-88-gaff3ab9
At creation version: 2019.1210.2731-v1.6.0-88-gaff3ab9
Metadata format:     2
Content checking:    size & timestamp
Data compression:    deactivated
Repository mode:     simple
Number of branches:  2
✔ File tree is unchanged                                             
 

Now we switch back to our initial branch:

In [57]:
!sos switch 0

MOD rese/sunamount.coco
/SOS/ Switched to branch 'master' b0/r01
[EXIT]


or use:

In [58]:
!sos switch master

No changes to current file tree
/SOS/ Switched to branch 'master' b0/r01
[EXIT]


You can see that the switch modified the source file that we modified back to its original state.
We might now decide that we even want to go back to the original project state on the trunk branch.
This rollback is also performed by the switch command:

In [59]:
!sos switch /0

MOD ./make.bat
/SOS/ Switched to branch 'master' b0/r00
[EXIT]


This will restore the current branch's initial state, which - in case of the first branch - is the state where we went offline from.

When you call `sos changes` and `sos status`, you'll see a major difference to other VCS and something stemming from SOS's simple data model:

In [60]:
!sos changes
!echo
!sos status

MOD ./make.bat <older than previously committed>
[EXIT]

/SOS/ Offline repository status
Repository root:     /home/ash/Desktop/all/projects/sos/docs/realestate-sunamount
Underlying VCS root: /home/ash/Desktop/all/projects/sos/docs/realestate-sunamount
Underlying VCS type: git
Installation path:   /home/ash/Desktop/all/projects/sos
Current SOS version: 2019.1210.2731-v1.6.0-88-gaff3ab9
At creation version: 2019.1210.2731-v1.6.0-88-gaff3ab9
Metadata format:     2
Content checking:    size & timestamp
Data compression:    deactivated
Repository mode:     simple
Number of branches:  2
✖ File tree has changes                                              
  * b0 'master' @2019-02-10 17:53:39 (modified) with 2 commits (median 7 each), using 0.13 MiB (+1.994% SOS overhead)
    b1          @2019-02-10 17:53:59 (modified) with 2 commits (median 7 each), using 0.15 MiB (+1.794% SOS overhead). Last comment: 'Moved import statement'
[EXIT]


You can see that the `make.bat` file is recognized as modified, also that SOS determined it to be older than a recent commit (by file modification timestamp).
The reason for marking the file as modified is that SOS compares the file tree always against the latest revision of the current branch.
There is no going back to a previous revision and forgetting about the relative future commits like Git can (by resetting its index).
The rationale behind it is that most work is sequential.

If we were to make a change now - from the file tree at b0/r0, it would be added as revision 2 of branch 0.

Before committing anything now, go ahead and change `rese/sunamount.coco` to import the `logging` module right before the `import time` statement:

In [61]:
with open("rese/sunamount.coco", "r") as fd: lines = fd.read().split("\n")
lines = lines[:11] + ["import logging"] + lines[11:]
with open("rese/sunamount.coco", "w") as fd: fd.write("\n".join(lines))

!sos diff

DIF ./make.bat  <LF>
old 0 |@echo off...|
now 0 |@echo off|

DIF rese/sunamount.coco  <LF>
+++ 11 |import logging|

[EXIT]


The rollback of the make file is marked as a change, and the added import is shown.
We continue and commit:

In [62]:
!sos commit "Added import statement"
!echo
!sos log

MOD ./make.bat <older than previously committed>
MOD rese/sunamount.coco
Processing speed was 4.44 MiB/s.
/SOS/ Created new revision r02 'Added import statement' (+00/-00/±02/⇌00) summing 16.31 KiB in 2 files (2.12% SOS overhead)
[EXIT]

/SOS/ Offline commit history of branch 'master'
    r0 @2019-02-10 17:53:39 (+14/-00/±00/T00) |Offline repository created on 2019-02-10 17:53:39|
    r1 @2019-02-10 17:53:51 (+00/-00/±01/T01) ||
  ➙ r2 @2019-02-10 17:54:27 (+00/-00/±02/T02) |Added import statement|
[EXIT]


When we compare the current file tree state with the initial revision, we see that there is no change except for the added `import logging`. We have rolled back the change from revision 0 to 1 and commited that reversal to revision 2:

In [63]:
!sos diff /0

DIF rese/sunamount.coco  <LF>
+++ 11 |import logging|

[EXIT]


## Merging ##
Ok, we covered branching and rollback.
Now we go to the royal discipline - merging changes from other branches or revisions into the file tree.

If you want to compare and replay changes from another branch into the current file tree, you would first check what the differences are:

In [64]:
!sos diff 1

DIF ./make.bat  <LF>
old 0 |@echo off...|
now 0 |@echo off|

DIF rese/sunamount.coco  <LF>
--- 008 |import sys|

+++ 012 |import logging|

+++ 275 |  import sys|

[EXIT]


Then we proceed to integrate everything differeing in branch 1 into our current master branch (branch 0):

In [65]:
!sos update 1
!echo
!sos diff

Modifications:
MRG ./make.bat
MRG rese/sunamount.coco
/SOS/ Integrated changes from 'b1/r01' into file tree
[EXIT]

DIF ./make.bat  <LF>
old 0 |@echo off|
now 0 |@echo off...|

DIF rese/sunamount.coco  <LF>
+++ 008 |import sys|

--- 012 |import logging|

--- 275 |  import sys|

[EXIT]


OK, let's review that. Here, `sos upgrade` does nothing different than what a `sos switch` would do, but doesn't switch the current branch - everything is just replayed on the file tree.
This example is probably not very useful, since both branches look now exactly the same.

These changes have not been committed from the file system to the repository, but we can confirm that we have now the same state as branch 1:

In [66]:
!sos diff 1/

DIF ./make.bat <timestamp or newline>  <LF>
DIF rese/sunamount.coco <timestamp or newline>  <LF>
[EXIT]


As you can see, the files do not differ, except for the timestamps (in this case).

To make it worthwhile, you might want to consult the documentation for the `sos update` command:

In [67]:
!sos help update

/SOS/ Subversion Offline Solution

Working with files:
  update [<branch>][/<revision>]  Integrate work from another branch into the file tree without switching the branch
                                  Similarly to switch, this command updates the current file tree to the state of another revision, usually from another branch.
                                  In addition, it allows merging differing contents interactively or by rules specified through command-line switches.
                                  Default operation for files, lines and characters is add and remove (recreate other branch in file tree)

  Arguments:
    [<branch>][/<revision>]  Branch and/or revision

  Options:
    --add                                                                Only add new files (won't remove)
    --add-lines                                                          Only add inserted lines
    --ask                                                                Ask how to proceed wit

OK, we want to try the update command once more, but this time in a more useful and refined manner.

To do so, let's go back to where we were before:

In [68]:
!sos switch trunk

MOD ./make.bat
MOD rese/sunamount.coco
[EXIT File tree contains changes. Use --force to proceed.]


OK; SOS didn't let us switch as there are local modifications. In Git you would have to stash or remove the files first.

In [69]:
!sos switch trunk --force

MOD ./make.bat
MOD rese/sunamount.coco
/SOS/ Switched to branch 'master' b0/r02
[EXIT]


In [70]:
!sos diff 1/

DIF ./make.bat  <LF>
old 0 |@echo off...|
now 0 |@echo off|

DIF rese/sunamount.coco  <LF>
--- 008 |import sys|

+++ 012 |import logging|

+++ 275 |  import sys|

[EXIT]


This time we retry the `sos update` command in combination with file patterns like `--only code/*.py` and integrate the changes from branch 1 step by step for each file or file pattern separately:

In [74]:
!sos update 1/ --only ./make.bat --ask-lines

Modifications:
MRG ./make.bat
/SOS/ Integrated changes from 'b1/r01' into file tree
[EXIT]


In [73]:
!sos diff 1

DIF ./make.bat <timestamp or newline>  <LF>
DIF rese/sunamount.coco  <LF>
--- 008 |import sys|

+++ 012 |import logging|

+++ 275 |  import sys|

[EXIT]
